### PR TITLE
testcase/le_tc/compression: Testcase for compress_read APIs

### PR DESCRIPTION
**Pull request by Nishtha97**

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [2b9829bb4c3734f1da60cc4c3cb85a2b51f76741] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [2b9829bb4c3734f1da60cc4c3cb85a2b51f76741] - Code Rule Check OK.

File: apps/examples/testcase/le_tc/compression/tc_compress_read.c [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <debug.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <tinyara/config.h>
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/compression/tc_compress_read.c]: Please place the header file like below sequence.
1. tinyara/config.h
2. standard header : stdio.h, unistd.h, string.h and so on.
3. tinyara/xxx.h
4. absolute path header : "xxx.h"

File: apps/examples/testcase/le_tc/compression/tc_compress_read.c [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+#include <stdint.h>
+#include <unistd.h>
+#include <string.h>
+#include <tc_common.h>
+
+/****************************************************************************
+ * Private Declarations
+ ****************************************************************************/
+
+struct s_header *compression_header;
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/compression/tc_compress_read.c]: Please add static.

File: os/compression/compress_read.c [hunk 3]
```diff
@@ -56,7 +58,7 @@ static struct s_buffer buffers;
  *   Non-negative value on Success.
  *   Negative value on Failure.
  ****************************************************************************/
-static int compress_decompress_block(unsigned char *out_buffer, unsigned int *writesize, unsigned char *read_buffer, unsigned int *size, int index)
+int compress_decompress_block(unsigned char *out_buffer, unsigned int *writesize, unsigned char *read_buffer, unsigned int *size, int index)
```
Comment (jeongchanKim) [inline os/compression/compress_read.c]: Is this for calling this function from testcase...?
If yes, how about using testcase_drv for testing these functions?

File: apps/examples/testcase/le_tc/compression/tc_compress_read.c [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	ret = compress_init(filefd, 0, &filelen);
+	if (ret != OK) {
+		berr("Failed to read header for compressed binary : %d\n", ret);
+		return ret;
+	}
+
+	for (i = 0; i < compression_header->sections - 1; i++) {
+		size = compress_read_block(filefd, 0, buffers.read_buffer, i);
+		if (size < 0) {
+			bmdbg("Read for compressed block %d failed\n", i);
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/compression/tc_compress_read.c]: Why don't you add TC_ASSERT_XXX macro for checking this tc?

File: apps/examples/testcase/le_tc/compression/tc_compress_read.c (hunk 1 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/tc_compress_read.c]: Changes made

**Comment by seinfra** [comment]
Target : [72f2fb08cfb50007b804ae3257a10a1907fc4d1c] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [72f2fb08cfb50007b804ae3257a10a1907fc4d1c] - Code Rule Check OK.

File: os/compression/compress_read.c (hunk 3 above)
Comment (Nishtha97) [reply os/compression/compress_read.c]: Currently the testcase only handles the flat build case. Also I will look into how to use testcase_drv for this

File: apps/examples/testcase/le_tc/compression/Kconfig [hunk 5]
```diff
@@ -16,4 +16,10 @@ config TC_LZMA
 	default y
 	select LIBLZMA
 
+config TC_COMPRESS_READ
```
Comment (thapav) [inline apps/examples/testcase/le_tc/compression/Kconfig]: May be this config can be added under a  TC_COMPRESS_ALL config as follows:
if EXAMPLES_TESTCASE_COMPRESSION
config TC_COMPRESS_ALL
        bool "All"
        default y
        select TC_LZMA
        select TC_COMPRESS_READ

config TC_LZMA
....
....
config TC_COMPRESS_READ
bool "Compress_read_testcase"
	default n
	select LIBLZMA
	select COMPRESSED_BINARY
endif

So that all the compression/decompression testcases can be enabled at once.

File: os/compression/compress_read.c (hunk 3 above)
Comment (thapav) [reply os/compression/compress_read.c]: I agree with jeongchan, using testcase_drv is a better approach to test this API.

**Comment by seinfra** [comment]
Target : [36930f3fd95ec03a6794e30138cbb78feda4ab06] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [36930f3fd95ec03a6794e30138cbb78feda4ab06] - Code Rule Check (C++) OK.

File: apps/examples/testcase/le_tc/compression/tc_compress_read.c (hunk 4 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/tc_compress_read.c]: Changes made.

**Comment by seinfra** [comment]
Target : [18305e9f62ba50b9eaa7d3af0b8ff59318ac260f] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [18305e9f62ba50b9eaa7d3af0b8ff59318ac260f] - Code Rule Check Result:
os/kernel/init/os_start.c:554: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kernel/init/os_start.c:554: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by seinfra** [comment]
Target : [eb8ab9b8daf557c888a5372ffdca7a33ba90ad75] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [eb8ab9b8daf557c888a5372ffdca7a33ba90ad75] - Code Rule Check (C++) OK.

File: os/compression/compress_read.c (hunk 3 above)
Comment (Nishtha97) [reply os/compression/compress_read.c]: Made the changes

**Comment by seinfra** [comment]
Target : [35ddad2a084da60869dd88546bcfc17b00e95455] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [35ddad2a084da60869dd88546bcfc17b00e95455] - Code Rule Check (C++) OK.

File: apps/examples/testcase/le_tc/compression/tc_compress_read.c (hunk 2 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/tc_compress_read.c]: Made changes in this file. This declaration is not made now

File: apps/examples/testcase/le_tc/compression/Kconfig [hunk 6]
```diff
@@ -6,6 +6,7 @@
 menuconfig EXAMPLES_TESTCASE_COMPRESSION
 	bool "Compression TestCase Example"
 	default n
+	select COMPRESSION_TEST_DRV
```
Comment (amandeepc) [inline apps/examples/testcase/le_tc/compression/Kconfig]: This should be under TC_COMPRESS_READ.
I think this is needed only in that case.

File: os/drivers/testcase/compression_testcase_drv.c [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ ************************************************************************************/
+
+static int compression_test_drv_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+	int ret = -EINVAL;
+	/* Handle built-in ioctl commands */
+
+	switch (cmd) {
+	/* TESTFWIOC_DRIVER_ANALOG - Run the test case for /os/driver/analog module
```
Comment (amandeepc) [inline os/drivers/testcase/compression_testcase_drv.c]: Proper indentation here.

File: os/include/tinyara/binfmt/compression/compress_read.h [hunk 8]
```diff
@@ -80,4 +80,30 @@ int compress_init(int filfd, uint16_t offset, off_t *filelen);
...
+ * Description:
+ *   Decompress block in 'read_buffer' of readsize into 'out_buffer' of writesize
+ *
+ * Returned Value:
+ *   Non-negative value on Success.
+ *   Negative value on Failure.
+ ****************************************************************************/
+int compress_decompress_block(unsigned char *out_buffer, unsigned int *writesize, unsigned char *read_buffer, unsigned int *size, int index);
+
+
```
Comment (amandeepc) [inline os/include/tinyara/binfmt/compression/compress_read.h]: get rid of extra lines

**Comment by seinfra** [comment]
Target : [06e1af0ab599a6a300519742c31283b809f738d7] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [06e1af0ab599a6a300519742c31283b809f738d7] - Code Rule Check (C++) OK.

File: os/include/tinyara/binfmt/compression/compress_read.h (hunk 8 above)
Comment (Nishtha97) [reply os/include/tinyara/binfmt/compression/compress_read.h]: Done

File: os/drivers/testcase/compression_testcase_drv.c (hunk 7 above)
Comment (Nishtha97) [reply os/drivers/testcase/compression_testcase_drv.c]: Done

File: apps/examples/testcase/le_tc/compression/Kconfig (hunk 6 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/Kconfig]: Changes made

File: os/drivers/testcase/compression_testcase_drv.c [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+ * Name: compression_test_drv_register
+ *
+ * Description:
+ *   Register /dev/testcase2
+ *
+ ****************************************************************************/
+
+void compression_test_drv_register(void)
+{
+	(void)register_driver(COMPRESSION_TC_DRVPATH, &compression_test_drv_fops, 0666, NULL);
```
Comment (sunghan-chang) [inline os/drivers/testcase/compression_testcase_drv.c]: Could you let me know why you want to make a new driver instead of tc drv?

File: os/compression/compress_read.c [hunk 10]
```diff
@@ -216,7 +218,7 @@ static off_t compress_lseek_block(int filfd, uint16_t binary_header_size, int bl
  *   Number of bytes read into read_buffer on Success
  *   Negative value on Failure
  ****************************************************************************/
-static off_t compress_read_block(int filfd, uint16_t binary_header_size, FAR uint8_t *buf, int block_number)
+off_t compress_read_block(int filfd, uint16_t binary_header_size, FAR uint8_t *buf, int block_number)
```
Comment (sunghan-chang) [inline os/compression/compress_read.c]: In my opinion, we had not better make internal functions to public.

**Comment by sunghan-chang** [comment]
Title is too long. 
```
testcase/le_tc/compression: xxxx
```
is enough for me.

File: os/compression/compress_read.c [hunk 11]
```diff
@@ -39,8 +41,8 @@
  * Private Declarations
  ****************************************************************************/
 
-static struct s_header *compression_header;
-static struct s_buffer buffers;
+struct s_header *compression_header;
+struct s_buffer buffers;
```
[the remark this answers is not in the document]
Comment (Nishtha97) [reply os/compression/compress_read.c]: Made the required changes

File: os/compression/compress_read.c (hunk 10 above)
Comment (Nishtha97) [reply os/compression/compress_read.c]: These functions were made public as they are being called in the testcase file. Is there some other standard practice we use to handle this?

**Comment by seinfra** [comment]
Target : [0a45c363a4508bd4cdf14a39390aeedac7e34a83] - Code Rule Check Result:
os/drivers/testcase/test_compress_decompress.c:72: ERROR: [SPC_M_OPR] spaces required around that '=' (ctx:VxV)
os/drivers/testcase/test_compress_decompress.c:73: ERROR: [SPC_M_OPR] spaces required around that '=' (ctx:VxV)

**Comment by seinfra** [comment]
Target : [0a45c363a4508bd4cdf14a39390aeedac7e34a83] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [a22cf0308fc1523c48992ca63516b9d814004e34] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [a22cf0308fc1523c48992ca63516b9d814004e34] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [072cc9c4042009cfd099ef7cf1713e182e1ac3cd] - Code Rule Check Result:
os/include/tinyara/testcase_drv.h:159: ERROR: [IDT_M_TAB] please, no space before tabs

**Comment by seinfra** [comment]
Target : [072cc9c4042009cfd099ef7cf1713e182e1ac3cd] - Code Rule Check (C++) OK.

File: os/drivers/testcase/compression_testcase_drv.c (hunk 9 above)
Comment (Nishtha97) [reply os/drivers/testcase/compression_testcase_drv.c]: Made the changes

**Comment by seinfra** [comment]
Target : [e319f0811e1a1e7fff25e0b48a590fd5907e9733] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [e319f0811e1a1e7fff25e0b48a590fd5907e9733] - Code Rule Check Result:
os/drivers/testcase/test_compress_decompress.c:78: ERROR: [SPC_M_OPR] spaces required around that '=' (ctx:VxV)
os/drivers/testcase/test_compress_decompress.c:80: ERROR: [SPC_M_KWD] space required before the open brace '{'
os/drivers/testcase/test_compress_decompress.c:80: ERROR: [SPC_M_KWD] space required before the open parenthesis '('

**Comment by seinfra** [comment]
Target : [2bdd6ead455768bc35a84752724d18764d9cb136] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [2bdd6ead455768bc35a84752724d18764d9cb136] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [824898c240495b8d13c3911cd6ac7275ec9850e4] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [824898c240495b8d13c3911cd6ac7275ec9850e4] - Code Rule Check OK.

File: os/compression/compress_read.c (hunk 10 above)
Comment (Nishtha97) [reply os/compression/compress_read.c]: Reverted the changes. Now these functions are no longer public functions.

File: apps/examples/testcase/le_tc/compression/compression_tc_main.c [hunk 12]
```diff
@@ -22,17 +22,41 @@
 
 #include <tinyara/config.h>
 #include <stdio.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <tinyara/testcase_drv.h>
+#include <tinyara/fs/ioctl.h>
+#include <sys/types.h>
+#include <sys/stat.h>
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/compression/compression_tc_main.c]: Why is this necessary?

File: apps/examples/testcase/le_tc/compression/compression_tc_main.c [hunk 13]
```diff
@@ -22,17 +22,41 @@
 
 #include <tinyara/config.h>
 #include <stdio.h>
+#include <errno.h>
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/compression/compression_tc_main.c]: Why is this necessary?

File: apps/examples/testcase/le_tc/compression/tc_compress_read.c [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+#include <tinyara/config.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <tc_common.h>
+#include <tinyara/testcase_drv.h>
+#include <tinyara/binfmt/compression/compress_read.h>
+#include "tc_internal.h"
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/compression/tc_compress_read.c]: Could you check all of headers are necessary?

File: os/compression/compress_read.c [hunk 15]
```diff
@@ -416,3 +416,14 @@ void compress_uninit(void)
 	kmm_free(compression_header);
 	compression_header = NULL;
 }
+
+struct s_header* get_compression_header(void)
+{
+	return compression_header;
+}
+
+struct s_buffer* get_buffers(void)
```
Comment (sunghan-chang) [inline os/compression/compress_read.c]: struct s_header *get_compression_header
struct s_buffer *get_buffers

File: os/drivers/testcase/test_compress_decompress.c [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+#include <fcntl.h>
+
+/****************************************************************************
+ * Private Declarations
+ ****************************************************************************/
+
+static struct s_header *compression_header;
+static struct s_buffer *buffers;
+
+FAR uint8_t *buffer;
```
Comment (sunghan-chang) [inline os/drivers/testcase/test_compress_decompress.c]: Please add "static" and remove "FAR".

Comment (sunghan-chang) [reply os/drivers/testcase/test_compress_decompress.c]: buffers and buffer are not good naming because I can't know what they are.

File: os/drivers/testcase/test_compress_decompress.c [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+
+	if (ret != OK) {
+		berr("Failed to read header for compressed binary : %d\n", ret);
+		return ret;
+	}
+
+	compression_header = get_compression_header();
+	buffers = get_buffers();
+
+	buffer = (FAR uint8_t *)malloc(2048 * sizeof(uint8_t));
```
Comment (sunghan-chang) [inline os/drivers/testcase/test_compress_decompress.c]: Where is freeing?

File: os/drivers/testcase/test_compress_decompress.c [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+
+	if (buffer == NULL) {
+		berr("Allocation of memory failed\n", i);
```
Comment (sunghan-chang) [inline os/drivers/testcase/test_compress_decompress.c]: Please check the argument, i.

File: os/drivers/testcase/test_compress_decompress.c [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+		return ERROR;
+	}
+
+
+	for (i = 0; i < (compression_header->sections - 1); i++) {
+		size = compress_read(filefd, 0, buffer, readsize, (i*2048));
```
Comment (sunghan-chang) [inline os/drivers/testcase/test_compress_decompress.c]: i * 2048

File: os/include/tinyara/binfmt/compression/compress_read.h [hunk 20]
```diff
@@ -80,4 +80,20 @@ int compress_init(int filfd, uint16_t offset, off_t *filelen);
...
+ ****************************************************************************/
+struct s_header* get_compression_header(void);
+
+/****************************************************************************
+ * Name: get_buffers
+ *
+ * Returned Value:
+ *   Address of the buffer
+ ****************************************************************************/
+struct s_buffer* get_buffers(void);
```
Comment (sunghan-chang) [inline os/include/tinyara/binfmt/compression/compress_read.h]: what buffers?

File: apps/examples/testcase/le_tc/compression/Kconfig [hunk 21]
```diff
@@ -16,4 +16,11 @@ config TC_LZMA
 	default y
 	select LIBLZMA
 
+config TC_COMPRESS_READ
+	bool "Testcase for Compression APIs"
+	default y
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/compression/Kconfig]: default n

File: apps/examples/testcase/le_tc/compression/tc_lzma.c [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+		arr[i] = 'A' + ((rand()) % 26);
+	}
+
+	/* CRC of the buffer initially */
+	ini_t = crc32(arr, SIZE_4KB);
+
+	ret = LzmaCompress(&out_buf[LZMA_PROPS_SIZE], &writesize, arr, SIZE_4KB, out_buf, &propsSize, 0, 1 << 13, -1, -1, -1, -1, 1);
+
+	TC_ASSERT_EQ("LzmaCompress", ret, SZ_OK);
+	/* If FAIL: Dynamic RAM size needed by Lzma is not enough */
```
[the remark this answers is not in the document]
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/tc_lzma.c]: Done

**Comment by seinfra** [comment]
Target : [29332210ab77a1f38490d935c708b965050e2d1f] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [29332210ab77a1f38490d935c708b965050e2d1f] - Code Rule Check (C++) OK.

File: apps/examples/testcase/le_tc/compression/compression_tc_main.c (hunk 12 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/compression_tc_main.c]: Removed the unnecessary header files.

File: apps/examples/testcase/le_tc/compression/tc_compress_read.c (hunk 14 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/tc_compress_read.c]: Done

File: os/drivers/testcase/test_compress_decompress.c (hunk 16 above)
Comment (Nishtha97) [reply os/drivers/testcase/test_compress_decompress.c]: Done. Changed the "buffer" name to "dst_buffer".

File: os/drivers/testcase/test_compress_decompress.c (hunk 18 above)
Comment (Nishtha97) [reply os/drivers/testcase/test_compress_decompress.c]: Changes made

File: os/include/tinyara/binfmt/compression/compress_read.h (hunk 20 above)
Comment (Nishtha97) [reply os/include/tinyara/binfmt/compression/compress_read.h]: According to the changes made, this function is not needed anymore and hence is removed.

File: apps/examples/testcase/le_tc/compression/Kconfig (hunk 21 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/Kconfig]: Done

**Comment by seinfra** [comment]
Target : [83e6d70981c34aadfd6b85b1ffc45ea27177cf10] - Code Rule Check (C++) OK.

File: apps/examples/testcase/le_tc/compression/compression_tc_main.c (hunk 13 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/compression_tc_main.c]: Removed the unnecessary header files.

File: os/compression/compress_read.c (hunk 15 above)
Comment (Nishtha97) [reply os/compression/compress_read.c]: Changes made.

File: os/drivers/testcase/test_compress_decompress.c (hunk 17 above)
Comment (Nishtha97) [reply os/drivers/testcase/test_compress_decompress.c]: Required changes made.

File: os/drivers/testcase/test_compress_decompress.c (hunk 19 above)
Comment (Nishtha97) [reply os/drivers/testcase/test_compress_decompress.c]: Done

**Comment by seinfra** [comment]
Target : [83e6d70981c34aadfd6b85b1ffc45ea27177cf10] - Code Rule Check OK.

File: os/include/tinyara/testcase_drv.h [hunk 23]
```diff
@@ -74,6 +74,7 @@
 #define TESTIOC_GROUP_REMOVECHILDREN_TEST      _TESTIOC(17)
 #endif
 #define TESTIOC_TASK_INIT_TEST                 _TESTIOC(18)
+#define TESTIOC_COMPRESSION_TEST	       _TESTIOC(19)
```
Comment (sunghan-chang) [inline os/include/tinyara/testcase_drv.h]: Sorry for another comments.
1. Could you align `_TESTIOC(10)`?
2. Could you leave the result logs in commit description?

**Comment by seinfra** [comment]
Target : [290173efd87782c7d770116f26a9294dd7786668] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [290173efd87782c7d770116f26a9294dd7786668] - Code Rule Check (C++) OK.

File: os/include/tinyara/testcase_drv.h (hunk 23 above)
Comment (Nishtha97) [reply os/include/tinyara/testcase_drv.h]: Changes made. Added the result logs in commit description.
